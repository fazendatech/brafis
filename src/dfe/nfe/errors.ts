import type { GetWebServiceUrlOptions } from "@nfe/types";

/**
 * @description Não foi localizado a url do web service.
 */
export class WebServiceNotFoundError extends Error {
  constructor({ service, uf, env, contingency }: GetWebServiceUrlOptions) {
    super(
      `Web service not found for ${service}/${uf} (${env}${contingency ? ", em contingência" : ""})`,
    );
    this.name = "WebServiceNotFoundError";
  }
}

/**
 * @description Falha ao fazer uma requisição de serviço.
 */
export class ServiceRequestError extends Error {
  public readonly xml: string;
  public readonly error: Error;

  constructor(error: Error, extra: { url: string; xml: string }) {
    super(`${error.message} URL: ${extra.url}`);
    this.name = "ServiceRequestError";
    this.stack = error.stack;
    this.error = error;
    this.xml = extra.xml;
  }
}
