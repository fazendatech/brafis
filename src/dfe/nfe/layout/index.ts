import { z } from "zod";
import { schemaNfeInfNfe } from "./groupA";

const schemaNfe = z
  .object({
    NFe: z.object({
      "@_xmlns": z.literal("http://www.portalfiscal.inf.br/nfe"),
      infNFe: schemaNfeInfNfe,
    }),
  })
  .describe("NFe (TAG Raiz)");

/**
 * @description Tipo que infere a estrutura da NFe.
 */
export type NfeLayout = z.infer<typeof schemaNfe>;

/**
 * @description Realiza a validação da NFe.
 *
 * @param {NfeLayout} nfe - Objeto NFe.
 *
 * @returns {NfeLayout} O objeto da NFe validado.
 * @throws {Zod.ZodError} Se a NFe não for válida.
 */
export function parseNfe(nfe: NfeLayout): NfeLayout {
  return schemaNfe.parse(nfe);
}
