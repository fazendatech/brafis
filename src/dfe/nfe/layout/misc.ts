import { z } from "zod";
import { ufCodeMap } from "@/ufCode";
import type { UF, UFCode } from "@/ufCode/types";
import type { NfeLayout } from ".";

// NOTE: Hack pro zod aceitar tipos customizados (https://stackoverflow.com/a/73825370)
const ufCodeList = Object.values(ufCodeMap);
const ufCodes: [UFCode, ...UFCode[]] = [ufCodeList[0], ...ufCodeList];
export const zUfCode = () => z.enum(ufCodes);

const ufList = Object.keys(ufCodeMap);
const ufs: [UF, ...UF[]] = [ufList[0], ...ufList];
export const zUf = () => z.enum(ufs);

// NOTE: NFe válida usada para testes
export const NFE_TEST_DATA: NfeLayout = {
  NFe: {
    "@_xmlns": "http://www.portalfiscal.inf.br/nfe",
    infNFe: {
      "@_versao": "4.00",
      "@_Id": "NFe12345678901234567890123456789012345678901235",
      ide: {
        cUF: "35",
        cNF: "00012345",
        natOp: "Venda",
        mod: "55",
        serie: "001",
        nNF: "123456",
        dhEmi: "2024-12-06T08:00:00Z",
        tpNF: "1",
        idDest: "1",
        cMunFG: "3550308",
        tpImp: "1",
        tpEmis: "1",
        cDV: "5",
        tpAmb: "1",
        finNFe: "1",
        indFinal: "1",
        indPres: "1",
        procEmi: "0",
        verProc: "1.0.0",
        NFref: [
          {
            refNFe: "12345678901234567890123456789012345678901234",
          },
          {
            refNF: {
              cUF: "35",
              AAMM: "2412",
              CNPJ: "12345678000195",
              mod: "01",
              serie: "001",
              nNF: "123456",
            },
          },
          {
            refNFP: {
              cUF: "35",
              AAMM: "2412",
              CPF: "11155599900",
              IE: "1234567890",
              mod: "04",
              serie: "001",
              nNF: "123456",
            },
          },
          {
            refCTe: "12345678901234567890123456789012345678901234",
          },
          {
            refECF: {
              mod: "2B",
              nECF: "123",
              nCOO: "123456",
            },
          },
        ],
      },
      emit: {
        CNPJ: "12345678000195",
        xNome: "Empresa Emitente LTDA",
        xFant: "Fantasia Emitente",
        enderEmit: {
          xLgr: "Rua Exemplo",
          nro: "123",
          xBairro: "Bairro Exemplo",
          cMun: "3550308",
          xMun: "São Paulo",
          UF: "SP",
          CEP: "01001000",
        },
        IE: "1234567890",
        CRT: "3",
      },
      avulsa: {
        CNPJ: "12345678000195",
        xOrgao: "Secretaria da Fazenda",
        matr: "123456789",
        xAgente: "Agente da Fazenda",
        fone: "11999999999",
        UF: "SP",
        nDAR: "123456789012345678901234567890123456789012345678901234567890",
        dEmi: "2024-12-06",
        vDAR: "500.00",
        repEmi: "0",
        dPag: "2024-12-06",
      },
      dest: {
        CPF: "11155599900",
        xNome: "Destinatário Exemplo",
        enderDest: {
          xLgr: "Avenida Exemplo",
          nro: "456",
          xBairro: "Outro Bairro",
          cMun: "3304557",
          xMun: "Rio de Janeiro",
          UF: "RJ",
          CEP: "20040030",
        },
        indIEDest: "9",
      },
      retirada: {
        CNPJ: "12345678000195",
        xLgr: "Rua Exemplo",
        nro: "123",
        xBairro: "Bairro Exemplo",
        cMun: "3550308",
        xMun: "São Paulo",
        UF: "SP",
      },
      entrega: {
        CNPJ: "12345678000195",
        xLgr: "Rua Exemplo",
        nro: "123",
        xBairro: "Bairro Exemplo",
        cMun: "3550308",
        xMun: "São Paulo",
        UF: "SP",
      },
      autXML: [
        {
          CNPJ: "12345678000195",
        },
      ],
      det: [
        {
          "@_nItem": "1",
          prod: {
            cProd: "001",
            cEAN: "SEM GTIN",
            xProd: "Produto Exemplo",
            NCM: "12345678",
            CFOP: "5102",
            uCom: "UN",
            qCom: "10.0000",
            vUnCom: "50.0000000000",
            vProd: "500.00",
            uTrib: "UN",
            qTrib: "10.0000",
            vUnTrib: "50.0000000000",
            indTot: "1",
            DI: [
              {
                nDI: "123456",
                dDI: "2024-12-06",
                xLocDesemb: "Local Desembaraço",
                UFDesemb: "SP",
                dDesemb: "2024-12-06",
                cExportador: "123456789",
                adi: [
                  {
                    nAdicao: "1",
                    nSeqAdic: "1",
                    cFabricante: "123456789",
                    vDescDI: "0.00",
                    nDraw: "12345678901",
                  },
                ],
                tpViaTransp: "7",
                tpIntermedio: "1",
              },
            ],
            detExport: [
              {
                nDraw: "12345678901",
                exportInd: {
                  nRE: "123456789012",
                  chNFe: "12345678901212345678901234567890123456789012",
                  qExport: "10.000",
                },
              },
            ],
            rastro: [
              {
                nLote: "1",
                qLote: "10.000",
                dFab: "2024-12-06",
                dVal: "2024-12-06",
                cAgreg: "123456789",
              },
            ],
          },
          imposto: {
            ICMS: {
              ICMS00: {
                orig: "0",
                CST: "00",
                modBC: "3",
                vBC: "500.00",
                pICMS: "18.00",
                vICMS: "90.00",
              },
            },
          },
        },
      ],
      total: {
        ICMSTot: {
          vBC: "500.00",
          vICMS: "90.00",
          vICMSDeson: "0.00",
          vFCP: "0.00",
          vBCST: "0.00",
          vST: "0.00",
          vFCPST: "0.00",
          vFCPSTRet: "0.00",
          vProd: "500.00",
          vFrete: "0.00",
          vSeg: "0.00",
          vDesc: "0.00",
          vII: "0.00",
          vIPI: "0.00",
          vIPIDevol: "0.00",
          vPIS: "0.00",
          vCOFINS: "0.00",
          vOutro: "0.00",
          vNF: "500.00",
        },
        ISSQNtot: {
          vServ: "0.00",
          vBC: "0.00",
          vISS: "0.00",
          vPIS: "0.00",
          vCOFINS: "0.00",
          dCompet: "2024-12-06",
          vDeducao: "0.00",
          vOutro: "0.00",
          vDescIncond: "0.00",
          vDescCond: "0.00",
          vISSRet: "0.00",
          cRegTrib: "5",
        },
        retTrib: {
          vRetPIS: "0.00",
          vRetCOFINS: "0.00",
          vRetCSLL: "0.00",
          vBCIRRF: "0.00",
          vIRRF: "0.00",
          vBCRetPrev: "0.00",
          vRetPrev: "0.00",
        },
      },
      transp: {
        modFrete: "1",
      },
      pag: {
        detPag: [
          {
            tPag: "01",
            vPag: "500.00",
          },
        ],
      },
      cobr: {
        dup: [
          {
            vDup: "500.00",
          },
        ],
      },
      infIntermed: {
        CNPJ: "00023456000177",
        idCadIntTran:
          "123456789012345678901234567890123456789012345678901234567890",
      },
      infAdic: {},
      infRespTec: {
        CNPJ: "12345678000195",
        xContato: "Responsável Técnico",
        email: "responsavel@empresa.com",
        fone: "11999999999",
      },
    },
  },
};
