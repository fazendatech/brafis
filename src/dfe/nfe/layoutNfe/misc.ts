import { z } from "zod";
import { ufCodeMap } from "@/ufCode";
import type { UF, UFCode } from "@/ufCode/types";

// NOTE: Hack pro zod aceitar tipos customizados (https://stackoverflow.com/a/73825370)
const ufCodeList = Object.values(ufCodeMap);
const ufCodes: [UFCode, ...UFCode[]] = [ufCodeList[0], ...ufCodeList];
export const zUfCode = () => z.enum(ufCodes);

const ufList = Object.keys(ufCodeMap);
const ufs: [UF, ...UF[]] = [ufList[0], ...ufList];
export const zUf = () => z.enum(ufs);
