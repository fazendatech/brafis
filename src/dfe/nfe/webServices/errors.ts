/**
 * @description Falha ao fazer uma requisição de serviço.
 */
export class NfeServiceRequestError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "NfeServiceRequestError";
  }
}
