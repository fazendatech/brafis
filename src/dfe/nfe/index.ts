import type { CertificateP12 } from "@/certificate";
import type {
  Environment,
  NfeRequestOptions,
  NfeStatusServicoRaw,
  NfeStatusServicoResponse,
  NfeConsultaCadastroOptions,
  NfeConsultaCadastroRaw,
  NfeConsultaCadastroResponse,
  UF,
  UFCode,
  WebService,
} from "@/dfe/nfe/types";
import { buildSoap, fetchWithTls, parseSoap } from "@/utils";
import { getWebServiceUrl } from "./webServiceUrls.ts";
import { loadNfeCa } from "./ca.ts";
import { getUfCode } from "./ufCode.ts";
import { ServiceRequestError } from "./errors.ts";
import { getInfoStatus } from "./infoStatus.ts";

/**
 * Opções para configurar um nfe web service.
 *
 * @interface NfeWebServicesOptions
 *
 * @property {UF} uf - A unidade federativa (estado) para a NFe.
 * @property {Environment} env - "producao" ou "homologacao".
 * @property {CertificateP12} certificate - O certificado P12 para autenticação.
 * @property {boolean} [contingency] - Se True, será usado o servidor de contingência.
 * @property {number} [timeout] - Timeout da requisição em ms, o valor padrão é 15 segundos.
 */
export interface NfeWebServicesOptions {
  uf: UF;
  env: Environment;
  certificate: CertificateP12;
  contingency?: boolean;
  timeout?: number;
}

export class NfeWebServices {
  private uf: UF;
  private env: Environment;
  private certificate: CertificateP12;
  private contingency: boolean;
  private timeout: number;
  private ca: string;
  private tpAmb: "1" | "2";
  private cUF: UFCode;
  private namespace: { "@_xmlns": string };

  constructor(options: NfeWebServicesOptions) {
    this.uf = options.uf;
    this.env = options.env;
    this.certificate = options.certificate;
    this.contingency = options.contingency ?? false;
    this.timeout = options.timeout ?? 15000;
    this.ca = "";

    this.tpAmb = options.env === "producao" ? "1" : "2";
    this.cUF = getUfCode(options.uf);

    this.namespace = {
      "@_xmlns": "http://www.portalfiscal.inf.br/nfe",
    };
  }

  private async getCa(): Promise<string> {
    if (!this.ca) {
      this.ca = await loadNfeCa();
    }
    return this.ca;
  }

  private getUrl(service: WebService): string {
    return getWebServiceUrl({
      service,
      uf: this.uf,
      env: this.env,
      contingency: this.contingency,
    });
  }

  private async request<Body, NfeRequestResponse>(
    url: string,
    options: NfeRequestOptions<Body>,
  ): Promise<NfeRequestResponse> {
    const { body, timeout } = options;
    const { cert, key } = this.certificate.asPem();

    return await fetchWithTls(url, {
      method: "POST",
      headers: { "Content-Type": "application/soap+xml; charset=utf-8" },
      body: buildSoap(body),
      tls: {
        cert,
        key,
        ca: await this.getCa(),
      },
      signal: AbortSignal.timeout(timeout),
    })
      .then(async (res) => {
        return await res.text();
      })
      .then((xml) => {
        return parseSoap<NfeRequestResponse>(xml);
      })
      .catch((error) => {
        throw new ServiceRequestError(error, { url, xml: buildSoap(body) });
      });
  }

  /**
   * @description Consulta do status do serviço prestado pelo Portal da Secretaria de Fazenda Estadual.
   *
   * @returns {Promise<NfeStatusServicoResponse>} O status do serviço.
   */
  async statusServico(): Promise<NfeStatusServicoResponse> {
    const reqResponse: { retConsStatServ: NfeStatusServicoRaw } =
      await this.request(this.getUrl("NfeStatusServico"), {
        timeout: this.timeout,
        body: {
          "@_xmlns":
            "http://www.portalfiscal.inf.br/nfe/wsdl/NFeStatusServico4",
          consStatServ: {
            ...this.namespace,
            "@_versao": "4.00",
            tpAmb: this.tpAmb,
            cUF: this.cUF,
            xServ: "STATUS",
          },
        },
      });

    return {
      status: getInfoStatus(reqResponse.retConsStatServ.cStat),
      description: reqResponse.retConsStatServ.xMotivo ?? "",
      raw: reqResponse.retConsStatServ,
    };
  }

  /**
   * @description Consulta o cadastro de contribuintes do ICMS da unidade federada.
   *
   * @param {NfeConsultaCadastroOptions} options - Opções para a consulta.
   * @returns {Promise<NfeConsultaCadastroResponse>} Informações sobre o cadastro do contribuinte.
   */
  async consultaCadastro(
    options: NfeConsultaCadastroOptions,
  ): Promise<NfeConsultaCadastroResponse> {
    const serviceResponse: { retConsCad: NfeConsultaCadastroRaw } =
      await this.request(this.getUrl("NfeConsultaCadastro"), {
        timeout: this.timeout,
        body: {
          "@_xmlns":
            "http://www.portalfiscal.inf.br/nfe/wsdl/CadConsultaCadastro4",
          ConsCad: {
            ...this.namespace,
            "@_versao": "2.00",
            infCons: { xServ: "CONS-CAD", UF: this.uf, ...options },
          },
        },
      });

    return {
      status: getInfoStatus(serviceResponse.retConsCad.infCons?.cStat),
      description: serviceResponse.retConsCad.infCons?.xMotivo ?? "",
      raw: serviceResponse.retConsCad,
    };
  }

  autorizacao() {
    throw new Error("Method not implemented.");
  }
}
