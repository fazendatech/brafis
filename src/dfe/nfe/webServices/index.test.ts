import {
  describe,
  test,
  expect,
  afterEach,
  beforeEach,
  spyOn,
  mock,
} from "bun:test";
import { mock as mockRequest, clearMocks } from "bun-bagel";
import { XMLBuilder } from "fast-xml-parser";

import { CertificateP12 } from "@/certificate";
import { NFE_TEST_DATA } from "@/dfe/nfe/layout/misc";
import { getWebServiceUrl } from "@/dfe/nfe/webServiceUrls";

import { NfeWebServices } from ".";
import { NfeServiceRequestError } from "./errors";
import { ZodError } from "zod";

function buildMockResponse<Obj>(obj: Obj): string {
  const xmlBuilder = new XMLBuilder({
    ignoreAttributes: false,
    attributeNamePrefix: "@_",
  });
  return xmlBuilder.build({
    "?xml": {
      "@_version": "1.0",
      "@_encoding": "UTF-8",
    },
    "soapenv:Envelope": {
      "@_xmlns:xsi": "http://www.w3.org/2001/XMLSchema-instance",
      "@_xmlns:xsd": "http://www.w3.org/2001/XMLSchema",
      "@_xmlns:soap": "http://www.w3.org/2003/05/soap-envelope",
      "soapenv:Body": {
        ...obj,
      },
    },
  });
}

function mockCertificateSignXml(signedObject: unknown) {
  mock.module("src/certificate/sign", () => ({
    signXml: () => signedObject,
  }));
}

describe("NfeWebServices", async () => {
  const certificate = new CertificateP12({
    pfx: new Uint8Array(),
    password: "",
  });

  const uf = "DF";
  const env = "homologacao";
  const service = new NfeWebServices({ uf, env, certificate });

  beforeEach(() => {
    spyOn(certificate, "asPem").mockReturnValueOnce({ cert: "", key: "" });
  });

  afterEach(() => {
    clearMocks();
  });

  describe("statusServico", () => {
    const url = getWebServiceUrl({ uf, env, service: "NfeStatusServico" });

    test("Returns status success", async () => {
      const mockResponse = {
        nfeResultMsg: {
          retConsStatServ: {
            tpAmb: "2",
            cStat: "107",
            xMotivo: "Servico em Operacao",
            cUF: "53",
          },
        },
      };
      mockRequest(url, {
        method: "POST",
        response: {
          data: buildMockResponse(mockResponse),
        },
      });

      const raw = mockResponse.nfeResultMsg.retConsStatServ;
      expect(await service.statusServico()).toMatchObject({
        status: "operando",
        description: raw.xMotivo,
        raw,
      });
    });

    test("Request throws NfeServiceRequestError on status not success", () => {
      mockRequest(url, {
        method: "POST",
        response: {
          status: 403,
        },
      });

      expect(() => service.statusServico()).toThrowError(
        // NOTE: Não existe atributo `MockResponse.statusText` no bun-bagel
        new NfeServiceRequestError(`403 (403) - ${url}`),
      );
    });

    test("Request throws NfeServiceRequestError on `nfeResultMsg` not present in response", () => {
      mockRequest(url, {
        method: "POST",
        response: {
          data: "request failed",
        },
      });

      expect(() => service.statusServico()).toThrowError(
        new NfeServiceRequestError(`URL: ${url}\nrequest failed`),
      );
    });

    test("Request throws error on timeout", () => {
      mockRequest(url, {
        method: "POST",
        throw: new Error("The operation timed out."),
      });

      expect(() => service.statusServico()).toThrowError(
        "The operation timed out.",
      );
    });

    test("Request throws NfeServiceRequestError on generic error", () => {
      mockRequest(url, {
        method: "POST",
        throw: new Error("An error occurred during the request."),
      });

      expect(() => service.statusServico()).toThrowError(
        new NfeServiceRequestError("An error occurred during the request."),
      );
    });
  });

  describe("consultaCadastro", () => {
    const url = getWebServiceUrl({ uf, env, service: "NfeConsultaCadastro" });

    test("Returns valid response", async () => {
      const mockResponse = {
        nfeResultMsg: {
          retConsCad: {
            infCons: {
              cStat: "111",
              xMotivo: "Consulta cadastro com uma ocorrência",
            },
          },
        },
      };
      mockRequest(url, {
        method: "POST",
        response: {
          data: buildMockResponse(mockResponse),
        },
      });

      const raw = mockResponse.nfeResultMsg.retConsCad;
      expect(await service.consultaCadastro({})).toMatchObject({
        status: "uma-ocorrencia",
        description: raw.infCons.xMotivo,
        raw,
      });
    });
  });

  describe("autorizacao", () => {
    const url = getWebServiceUrl({ uf, env, service: "NFeAutorizacao" });

    test("Returns valid response", async () => {
      mockCertificateSignXml({
        NFe: { infNFe: "mock NFe", Signature: "mock signature" },
      });

      mockRequest(url, {
        method: "POST",
        response: {
          data: buildMockResponse({
            nfeResultMsg: {
              retEnviNFe: {
                cStat: "104",
                xMotivo: "Lote processado",
                protNFe: {
                  "@_versao": "4.00",
                  infProt: {
                    tpAmb: "2",
                    verAplic: "SVRS202101041234",
                    chNFe: "0".repeat(44),
                    dhRecbto: "2021-01-04T12:34:56-03:00",
                    nProt: "123456789012345",
                    digVal: "123456789012345",
                    cStat: "100",
                    xMotivo: "Autorizado o uso da NF-e",
                  },
                },
              },
            },
          }),
        },
      });

      expect(
        await service.autorizacao({
          idLote: "1",
          nfe: NFE_TEST_DATA,
        }),
      ).toMatchSnapshot();
    });

    test("Webservice throws Zod.ZodError for an invalid NFe", () => {
      const nfe = NFE_TEST_DATA;
      nfe.NFe.infNFe.ide.cUF = "invalid";

      expect(() => service.autorizacao({ idLote: "1", nfe })).toThrowError(
        ZodError,
      );
    });
  });

  describe("inutilizacao", () => {
    const url = getWebServiceUrl({ uf, env, service: "NfeInutilizacao" });

    test("Returns valid response", async () => {
      mockCertificateSignXml({
        inutNfe: { infInut: "mock infInut", Signature: "mock signature" },
      });

      mockRequest(url, {
        method: "POST",
        response: {
          data: buildMockResponse({
            nfeResultMsg: {
              retInutNFe: {
                infInut: {
                  cStat: "102",
                  xMotivo: "Inutilização de número homologado",
                  cUF: "53",
                  nProt: "123456789",
                  dhRecbto: "2021-01-01T00:00:00-03:00",
                },
              },
            },
          }),
        },
      });

      expect(
        await service.inutilizacao({
          ano: "2021",
          cnpj: "12345678901230",
          mod: "55",
          serie: "1",
          nNfIni: "1",
          nNfFin: "10",
          xJust: "justificativa",
        }),
      ).toMatchSnapshot();
    });

    test("Webservice throws Zod.ZodError for an invalid CNPJ", () => {
      expect(() =>
        service.inutilizacao({
          ano: "2021",
          cnpj: "12345678901231",
          mod: "55",
          serie: "1",
          nNfIni: "1",
          nNfFin: "10",
          xJust: "justificativa",
        }),
      ).toThrowError(ZodError);
    });
  });

  describe("recepcaoEvento", () => {
    const url = getWebServiceUrl({ uf, env, service: "RecepcaoEvento" });

    test("Returns valid response", async () => {
      mockCertificateSignXml({
        evento: {
          infEvento: { detEvento: "Mock detEvento" },
          Signature: "mock signature",
        },
      });

      mockRequest(url, {
        method: "POST",
        response: {
          data: buildMockResponse({
            nfeResultMsg: {
              retEnvEvento: {
                "@_versao": "1.00",
                idLote: "1",
                tpAmb: "2",
                verAplic: "SVRS202101041234",
                cOrgao: "91",
                cStat: "128",
                xMotivo: "Lote processado",
                retEvento: {
                  "@_versao": "1.00",
                  infEvento: {
                    "@_Id": "ID123",
                    tpEvento: "110110",
                    xEvento: "Carta de Correção aplicada",
                    cStat: "135",
                    xMotivo: "Evento registrado e vinculado a NF-e",
                    chNFe: "0".repeat(44),
                    dhRegEvento: "2025-01-01T00:00:00-03:00",
                    nProt: "123456789",
                  },
                },
              },
            },
          }),
        },
      });

      expect(
        await service.recepcaoEvento({
          idLote: "1",
          CNPJ: "12345678901230",
          dhEvento: "2025-01-01T00:00:00-03:00",
          nSeqEvento: "1",
          chNFe: "0".repeat(44),
          detEvento: {
            descEvento: "Carta de Correção",
            xCorrecao: "correção",
          },
        }),
      ).toMatchSnapshot();
    });

    test("Webservice throws Zod.ZodError for an invalid CPF", () => {
      expect(() =>
        service.recepcaoEvento({
          idLote: "1",
          CPF: "12345678999",
          dhEvento: "2025-01-01T00:00:00-03:00",
          nSeqEvento: "1",
          chNFe: "0".repeat(44),
          detEvento: {
            descEvento: "Carta de Correção",
            xCorrecao: "correção",
          },
        }),
      ).toThrowError(ZodError);
    });

    test("Webservice throws Zod.ZodError for an invalid CNPJ", () => {
      expect(() =>
        service.recepcaoEvento({
          idLote: "1",
          CNPJ: "12345678901299",
          dhEvento: "2025-01-01T00:00:00-03:00",
          nSeqEvento: "1",
          chNFe: "0".repeat(44),
          detEvento: {
            descEvento: "Carta de Correção",
            xCorrecao: "correção",
          },
        }),
      ).toThrowError(ZodError);
    });
  });

  describe("consultaProtocolo", () => {
    const url = getWebServiceUrl({ uf, env, service: "NfeConsultaProtocolo" });

    test("Returns valid response", async () => {
      mockRequest(url, {
        method: "POST",
        response: {
          data: buildMockResponse({
            nfeResultMsg: {
              retConsSitNFe: {
                tpAmb: "2",
                cStat: "100",
                xMotivo: "Autorizado o uso da NF-e",
                cUF: "53",
                chNFe: "0".repeat(44),
                protNFe: {
                  infProt: {
                    tpAmb: "2",
                    chNFe: "0".repeat(44),
                    nProt: "123456789012345",
                    digVal: "123456789012345",
                    cStat: "100",
                    xMotivo: "Autorizado o uso da NF-e",
                  },
                },
              },
            },
          }),
        },
      });
      expect(
        await service.consultaProtocolo({ chNFe: "0".repeat(44) }),
      ).toMatchSnapshot();
    });
  });
});
