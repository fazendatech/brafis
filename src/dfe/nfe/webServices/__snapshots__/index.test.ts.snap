// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`NfeWebServices autorizacao Returns valid response 1`] = `
{
  "nfeResultMsg": {
    "retEnviNFe": {
      "cStat": "104",
      "protNFe": {
        "@_versao": "4.00",
        "infProt": {
          "cStat": "100",
          "chNFe": "00000000000000000000000000000000000000000000",
          "dhRecbto": "2021-01-04T12:34:56-03:00",
          "digVal": "123456789012345",
          "nProt": "123456789012345",
          "tpAmb": "2",
          "verAplic": "SVRS202101041234",
          "xMotivo": "Autorizado o uso da NF-e",
        },
      },
      "xMotivo": "Lote processado",
    },
  },
  "xml": "<?xml version="1.0" encoding="UTF-8"?><nfeProc xmlns="http://www.portalfiscal.inf.br/nfe" versao="4.00"><NFe><infNFe>mock NFe</infNFe><Signature>mock signature</Signature></NFe><protNFe versao="4.00"><infProt><tpAmb>2</tpAmb><verAplic>SVRS202101041234</verAplic><chNFe>00000000000000000000000000000000000000000000</chNFe><dhRecbto>2021-01-04T12:34:56-03:00</dhRecbto><nProt>123456789012345</nProt><digVal>123456789012345</digVal><cStat>100</cStat><xMotivo>Autorizado o uso da NF-e</xMotivo></infProt></protNFe></nfeProc>",
}
`;

exports[`NfeWebServices inutilizacao Returns valid response 1`] = `
{
  "nfeResultMsg": {
    "retInutNFe": {
      "infInut": {
        "cStat": "102",
        "cUF": "53",
        "dhRecbto": "2021-01-01T00:00:00-03:00",
        "nProt": "123456789",
        "xMotivo": "Inutilização de número homologado",
      },
    },
  },
}
`;

exports[`NfeWebServices recepcaoEvento Returns valid response 1`] = `
{
  "nfeResultMsg": {
    "retEnvEvento": {
      "@_versao": "1.00",
      "cOrgao": "91",
      "cStat": "128",
      "idLote": "1",
      "retEvento": {
        "@_versao": "1.00",
        "infEvento": {
          "@_Id": "ID123",
          "cStat": "135",
          "chNFe": "00000000000000000000000000000000000000000000",
          "dhRegEvento": "2025-01-01T00:00:00-03:00",
          "nProt": "123456789",
          "tpEvento": "110110",
          "xEvento": "Carta de Correção aplicada",
          "xMotivo": "Evento registrado e vinculado a NF-e",
        },
      },
      "tpAmb": "2",
      "verAplic": "SVRS202101041234",
      "xMotivo": "Lote processado",
    },
  },
  "xml": "<?xml version="1.0" encoding="UTF-8"?><procEventoNFe xmlns="http://www.portalfiscal.inf.br/nfe" versao="1.00"><evento versao="1.00"><infEvento><detEvento>Mock detEvento</detEvento></infEvento><Signature>mock signature</Signature></evento><retEvento versao="1.00"><infEvento Id="ID123"><tpEvento>110110</tpEvento><xEvento>Carta de Correção aplicada</xEvento><cStat>135</cStat><xMotivo>Evento registrado e vinculado a NF-e</xMotivo><chNFe>00000000000000000000000000000000000000000000</chNFe><dhRegEvento>2025-01-01T00:00:00-03:00</dhRegEvento><nProt>123456789</nProt></infEvento></retEvento></procEventoNFe>",
}
`;

exports[`NfeWebServices consultaProtocolo Returns valid response 1`] = `
{
  "nfeResultMsg": {
    "retConsSitNFe": {
      "cStat": "100",
      "cUF": "53",
      "chNFe": "00000000000000000000000000000000000000000000",
      "procEventoNFe": [
        {
          "evento": "mock evento",
          "retEvento": "mock ret",
        },
      ],
      "protNFe": {
        "infProt": {
          "cStat": "100",
          "chNFe": "00000000000000000000000000000000000000000000",
          "digVal": "123456789012345",
          "nProt": "123456789012345",
          "tpAmb": "2",
          "xMotivo": "Autorizado o uso da NF-e",
        },
      },
      "tpAmb": "2",
      "xMotivo": "Autorizado o uso da NF-e",
    },
  },
}
`;

exports[`NfeWebServices distribuicaoDfe Returns valid response 1`] = `
{
  "nfeDistDFeInteresseResponse": {
    "nfeDistDFeInteresseResult": {
      "retDistDFeInt": {
        "cStat": "138",
        "dhResp": "2025-01-24T08:00:00-03:00",
        "loteDistDFeInt": {
          "@_NSU": "0",
          "@_schema": "resNFe_v1.00.xsd",
          "docZip": [
            "NFe compactada no padrão gZip",
          ],
        },
        "maxNSU": "0",
        "tpAmb": "2",
        "ultNSU": "0",
        "xMotivo": "Documento localizado",
      },
    },
  },
}
`;
